flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableResultImpl.java

```diff
@@ -0,0 +1,334 @@
+	}
+
+	@Override
+	public TableSchema getTableSchema() {
+		return tableSchema;
+	}
+
+	@Override
+	public ResultKind getResultKind() {
+		return resultKind;
+	}
+
+	@Override
+	public Iterator<Row> collect() {
+		return data;
+	}
+
+	@Override
+	public void print() {
+		Iterator<Row> it = collect();
+		doPrint(it);
+	}
+
+	public static Builder builder() {
+		return new Builder();
+	}
+
+	/**
+	 * Builder for creating a {@link TableResultImpl}.
+	 */
+	public static class Builder {
+		private JobClient jobClient = null;
+		private TableSchema tableSchema = null;
+		private ResultKind resultKind = null;
+		private Iterator<Row> data = null;
+
+		private Builder() {
+		}
+
+		/**
+		 * Specifies job client which associates the submitted Flink job.
+		 *
+		 * @param jobClient a {@link JobClient} for the submitted Flink job.
+		 */
+		public Builder jobClient(JobClient jobClient) {
+			this.jobClient = jobClient;
+			return this;
+		}
+
+		/**
+		 * Specifies table schema of the execution result.
+		 *
+		 * @param tableSchema a {@link TableSchema} for the execution result.
+		 */
+		public Builder tableSchema(TableSchema tableSchema) {
+			Preconditions.checkNotNull(tableSchema, "tableSchema should not be null");
+			this.tableSchema = tableSchema;
+			return this;
+		}
+
+		/**
+		 * Specifies result kind of the execution result.
+		 *
+		 * @param resultKind a {@link ResultKind} for the execution result.
+		 */
+		public Builder resultKind(ResultKind resultKind) {
+			Preconditions.checkNotNull(resultKind, "resultKind should not be null");
+			this.resultKind = resultKind;
+			return this;
+		}
+
+		/**
+		 * Specifies an row iterator as the execution result .
+		 *
+		 * @param rowIterator a row iterator as the execution result.
+		 */
+		public Builder data(Iterator<Row> rowIterator) {
+			Preconditions.checkNotNull(rowIterator, "rowIterator should not be null");
+			this.data = rowIterator;
+			return this;
+		}
+
+		/**
+		 * Specifies an row list as the execution result .
+		 *
+		 * @param rowList a row list as the execution result.
+		 */
+		public Builder data(List<Row> rowList) {
+			Preconditions.checkNotNull(rowList, "listRows should not be null");
+			this.data = rowList.iterator();
+			return this;
+		}
+
+		/**
+		 * Returns a {@link TableResult} instance.
+		 */
+		public TableResult build() {
+			return new TableResultImpl(jobClient, tableSchema, resultKind, data);
+		}
+	}
+
+	/**
+	 * Displays the result in a tableau form.
+	 * For example:
+	 * +-------------+---------+-------------+
+	 * | boolean_col | int_col | varchar_col |
+	 * +-------------+---------+-------------+
+	 * |        true |       1 |         abc |
+	 * |       false |       2 |         def |
+	 * |      (NULL) |  (NULL) |      (NULL) |
+	 * +-------------+---------+-------------+
+	 * 3 row(s) in result
+	 *
+	 * <p>Changelog is not supported until FLINK-16998 is finished.
+	 */
+	private void doPrint(Iterator<Row> it) {
```

Comment:
Find a way to reuse codes instead of copying them?

Reply:
Maybe we can move this to `flink-table-common`